### PR TITLE
[FLINK-11560] Translate "Flink Applications" page into Chinese

### PR DESCRIPTION
**Pull request by crazyzhou**
[FLINK-11560] Translate "Flink Applications" page into Chinese

### CODE REVIEW

File: flink-applications.zh.md [hunk 1]
```diff
@@ -16,71 +16,72 @@ title: "Apache Flink 是什么？"
...
-* **Exactly-once state consistency**: Flink's checkpointing and recovery algorithms guarantee the consistency of application state in case of a failure. Hence, failures are transparently handled and do not affect the correctness of an application.
-* **Very Large State**: Flink is able to maintain application state of several terabytes in size due to its asynchronous and incremental checkpoint algorithm.
-* **Scalable Applications**: Flink supports scaling of stateful applications by redistributing the state to more or fewer workers.
+* **多种状态原语支持**：Flink 为多种不同的数据结构提供了相对应的状态存储原语，例如基本数据类型（value），列表（list）以及映射（map）。开发者可以基于处理函数对状态的访问方式，选择最高效、最适合的状态原语。
+* **插件式的State Backend**：State Backend 负责管理和在 checkpoint 时存储应用程序的状态。Flink 支持多种 state backend，包括内存内存储以及 [RocksDB](https://rocksdb.org/)。RocksDB 是一种高效的嵌入式、持久化的键值存储。Flink 同时支持插件式的自定义 state backend 进行状态的存储。
+* **精确一次处理的状态一致性**：Flink 的 checkpoint 和故障恢复算法保证了故障发生后应用状态的一致性。因而，Flink 能够在不影响应用程序的正确性的条件下，透明地处理可能发生的故障。
```
Comment (klion26) [inline flink-applications.zh.md]: `精确一次处理的状态一致性` --> `精确一次语义`
`因而，Flink 能够在不影响应用程序的正确性的条件下，透明地处理可能发生的故障。` --> `因此，Flink 能够在应用程序发生故障时，对应用程序透明，不造成正确性的影响。`

File: flink-applications.zh.md [hunk 2]
```diff
@@ -16,71 +16,72 @@ title: "Apache Flink 是什么？"
...
+* **数据量非常大的状态**：Flink 能够利用其异步以及增量式的 checkpoint 算法，存储数 TB 级别的应用状态。
```
Comment (klion26) [inline flink-applications.zh.md]: `数据量非常大的状态` --> `大数据量状态`

File: flink-applications.zh.md [hunk 3]
```diff
@@ -16,71 +16,72 @@ title: "Apache Flink 是什么？"
...
-* **Pluggable State Backends**: Application state is managed in and checkpointed by a pluggable state backend. Flink features different state backends that store state in memory or in [RocksDB](https://rocksdb.org/), an efficient embedded on-disk data store. Custom state backends can be plugged in as well.
-* **Exactly-once state consistency**: Flink's checkpointing and recovery algorithms guarantee the consistency of application state in case of a failure. Hence, failures are transparently handled and do not affect the correctness of an application.
-* **Very Large State**: Flink is able to maintain application state of several terabytes in size due to its asynchronous and incremental checkpoint algorithm.
-* **Scalable Applications**: Flink supports scaling of stateful applications by redistributing the state to more or fewer workers.
+* **多种状态原语支持**：Flink 为多种不同的数据结构提供了相对应的状态存储原语，例如基本数据类型（value），列表（list）以及映射（map）。开发者可以基于处理函数对状态的访问方式，选择最高效、最适合的状态原语。
```
[the remark this answers is not in the document]
Comment (crazyzhou) [reply flink-applications.zh.md]: 这里primitive意译成基本类型感觉更好一点，这样也解决了后面的primitive(n.)翻译

File: flink-applications.zh.md [hunk 4]
```diff
@@ -16,71 +16,72 @@ title: "Apache Flink 是什么？"
...
+* **插件式的State Backend**：State Backend 负责管理和在 checkpoint 时存储应用程序的状态。Flink 支持多种 state backend，包括内存内存储以及 [RocksDB](https://rocksdb.org/)。RocksDB 是一种高效的嵌入式、持久化的键值存储。Flink 同时支持插件式的自定义 state backend 进行状态的存储。
```
[the remark this answers is not in the document]
Comment (crazyzhou) [reply flink-applications.zh.md]: 我觉得自定义的state backend是以插件形式支持的，“可插拔”感觉更像硬件的描述。

File: flink-applications.zh.md [hunk 5]
```diff
@@ -16,70 +16,71 @@ title: "Apache Flink 是什么？"
...
-* **Late Data Handling**: When processing streams in event-time mode with watermarks, it can happen that a computation has been completed before all associated events have arrived. Such events are called late events. Flink features multiple options to handle late events, such as rerouting them via side outputs and updating previously completed results.
-* **Processing-time Mode**: In addition to its event-time mode, Flink also supports processing-time semantics which performs computations as triggered by the wall-clock time of the processing machine. The processing-time mode can be suitable for certain applications with strict low-latency requirements that can tolerate approximate results.
+* **事件时间模式**：使用事件时间语义的流处理应用根据事件本身自带的时间戳进行结果的计算。因此，无论处理的是历史记录的事件还是实时的事件，事件时间模式的处理总能保证结果的准确性和一致性。
+* **Watermark支持**：Flink 引入了 watermark 的概念，用以推断事件时间应用的时间。Watermark 也是一种平衡处理延时和完整性的灵活机制。
+* **迟到数据处理**：当以带有 watermark 的事件时间模式处理数据流时，会发生在所有相关数据到达之前计算就已经完成的情况。这样的事件被称为迟到事件。Flink 提供了多种处理迟到数据的选项，例如将这些数据重定向到旁路输出（side output）或者更新之前完成计算的结果。
```
Comment (klion26) [inline flink-applications.zh.md]: seems like `会发生在所有相关数据到达之前计算就已经完成的情况` can be improved. maybe `在计算完成之后仍会有数据到来`

File: flink-applications.zh.md [hunk 6]
```diff
@@ -16,70 +16,71 @@ title: "Apache Flink 是什么？"
...
+* **处理时间模式**：除了事件时间模式，Flink 还支持处理时间语义。处理时间模式根据处理引擎的机器时钟触发计算，一般适用于有着严格的低延迟需求，并且能够容忍近似结果的流处理应用。
```
Comment (klion26) [inline flink-applications.zh.md]: Do we need to translate `processing-time`?

File: flink-applications.zh.md [hunk 7]
```diff
@@ -16,70 +16,71 @@ title: "Apache Flink 是什么？"
...
+下文中，我们将简要描述每一种 API 及其应用，并提供相关的代码示例。
 
-### The ProcessFunctions
+### ProcessFunction
 
-[ProcessFunctions](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/operators/process_function.html) are the most expressive function interfaces that Flink offers. Flink provides ProcessFunctions to process individual events from one or two input streams or events that were grouped in a window. ProcessFunctions provide fine-grained control over time and state. A ProcessFunction can arbitrarily modify its state and register timers that will trigger a callback function in the future. Hence, ProcessFunctions can implement complex per-event business logic as required for many [stateful event-driven applications]({{ site.baseurl }}/usecases.html#eventDrivenApps).
+[ProcessFunction](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/operators/process_function.html) 是 Flink 所提供的最具表达力的接口。ProcessFunction 可以处理一或两条输入数据流中的单个事件或者归入一个特定窗口内的多个事件。它提供了对于时间和状态的细粒度控制。开发者可以在其中任意地修改状态，也能够注册计时器用以在未来的某一时刻触发回调函数。因此，你可以利用 ProcessFunction 实现许多[有状态的事件驱动应用]({{ site.baseurl }}/zh/usecases.html#eventDrivenApps)所需要的基于单个事件的复杂业务逻辑。
```
Comment (klion26) [inline flink-applications.zh.md]: I'm not very sure whether we should use `计时器` or `定时器` here for `timer`. or do we need to translate `timer` here?

File: flink-applications.zh.md [hunk 8]
```diff
@@ -16,70 +16,71 @@ title: "Apache Flink 是什么？"
...
 
-The following example shows a `KeyedProcessFunction` that operates on a `KeyedStream` and matches `START` and `END` events. When a `START` event is received, the function remembers its timestamp in state and registers a timer in four hours. If an `END` event is received before the timer fires, the function computes the duration between `END` and `START` event, clears the state, and returns the value. Otherwise, the timer just fires and clears the state.
+下面的代码示例展示了如何在 `KeyedStream` 上利用 `KeyedProcessFunction` 对标记为 `START` 和 `END` 的事件进行处理。当收到 `START` 事件时，处理函数会记录其时间戳，并且注册一个时长4小时的计时器。如果在计时器结束之前收到 `END` 事件，处理函数会计算其与上一个 `START` 事件的时间间隔，清空状态并将计算结果返回。否则，计时器结束，只进行状态的清空。
```
Comment (klion26) [inline flink-applications.zh.md]: `计时器结束，只进行状态的清空` --> `计时器结束，并清空状态`

File: flink-applications.zh.md [hunk 9]
```diff
@@ -133,59 +134,59 @@ public static class StartEndDuration
...
 {% endhighlight %}
 
-The example illustrates the expressive power of the `KeyedProcessFunction` but also highlights that it is a rather verbose interface.
+这个例子充分展现了 `KeyedProcessFunction` 强大的表达力，也因此是一个实现相当复杂的接口。
 
-### The DataStream API
+### DataStream API
 
-The [DataStream API](https://ci.apache.org/projects/flink/flink-docs-stable/dev/datastream_api.html) provides primitives for many common stream processing operations, such as windowing, record-at-a-time transformations, and enriching events by querying an external data store. The DataStream API is available for Java and Scala and is based on functions, such as `map()`, `reduce()`, and `aggregate()`. Functions can be defined by extending interfaces or as Java or Scala lambda functions. 
+[DataStream API](https://ci.apache.org/projects/flink/flink-docs-stable/dev/datastream_api.html) 为许多通用的流处理操作提供了处理原语。这些操作包括窗口、逐条记录的转移操作，在处理事件时进行外部数据库查询等。DataStream API 支持 Java 和 Scala 语言，预先定义了例如`map()`、`reduce()`、`aggregate()` 的函数。你可以通过扩展实现预定义接口或使用 Java、Scala 的 lambda 表达式实现自定义的函数。
```
Comment (klion26) [inline flink-applications.zh.md]: `逐条记录的转移操作` --> `逐条记录的转换操作`
`aggregate()` 的函数` --> `aggregate()  等函数`
```suggestion
[DataStream API](https://ci.apache.org/projects/flink/flink-docs-stable/dev/datastream_api.html) 为许多通用的流处理操作提供了处理原语。这些操作包括窗口、逐条记录的转换操作，在处理事件时进行外部数据库查询等。DataStream API 支持 Java 和 Scala 语言，预先定义了例如`map()`、`reduce()`、`aggregate()` 等函数。你可以通过扩展实现预定义接口或使用 Java、Scala 的 lambda 表达式实现自定义的函数。 
```

File: flink-applications.zh.md [hunk 10]
```diff
@@ -133,59 +134,59 @@ public static class StartEndDuration
...
-// a stream of website clicks
+// 网站点击 Click 的数据流
 DataStream<Click> clicks = ...
 
 DataStream<Tuple2<String, Long>> result = clicks
-  // project clicks to userId and add a 1 for counting
+  // 将网站点击映射为(userId, 1)以便计数
   .map(
-    // define function by implementing the MapFunction interface.
+    // 实现MapFunction接口定义函数
```
Comment (klion26) [inline flink-applications.zh.md]:
```suggestion
    // 实现 MapFunction 接口定义函数
```

File: flink-applications.zh.md [hunk 11]
```diff
@@ -133,59 +134,59 @@ public static class StartEndDuration
...
+下面的代码示例展示了如何捕获会话时间范围内所有的点击流事件，并对每一次会话的点击量进行计数。
 
 {% highlight java %}
-// a stream of website clicks
+// 网站点击 Click 的数据流
 DataStream<Click> clicks = ...
 
 DataStream<Tuple2<String, Long>> result = clicks
-  // project clicks to userId and add a 1 for counting
+  // 将网站点击映射为(userId, 1)以便计数
```
Comment (klion26) [inline flink-applications.zh.md]:
```suggestion
  // 将网站点击映射为 (userId, 1) 以便计数
```

File: flink-applications.zh.md [hunk 12]
```diff
@@ -133,59 +134,59 @@ public static class StartEndDuration
...
-    // define function by implementing the MapFunction interface.
+    // 实现MapFunction接口定义函数
     new MapFunction<Click, Tuple2<String, Long>>() {
       @Override
       public Tuple2<String, Long> map(Click click) {
         return Tuple2.of(click.userId, 1L);
       }
     })
-  // key by userId (field 0)
+  // 以 userId (field 0)作为 key
```
Comment (klion26) [inline flink-applications.zh.md]:
```suggestion
  // 以 userId (field 0) 作为 key
```

File: flink-applications.zh.md [hunk 13]
```diff
@@ -133,59 +134,59 @@ public static class StartEndDuration
...
   .keyBy(0)
-  // define session window with 30 minute gap
+  // 定义30分钟超时时间的会话窗口
```
Comment (klion26) [inline flink-applications.zh.md]:
```suggestion
  // 定义 30 分钟超时的会话窗口
```

File: flink-applications.zh.md [hunk 14]
```diff
@@ -133,59 +134,59 @@ public static class StartEndDuration
...
 ~~~sql
 SELECT userId, COUNT(*)
 FROM clicks
 GROUP BY SESSION(clicktime, INTERVAL '30' MINUTE), userId
 ~~~
 
-## Libraries
+## 库
+
+Flink具有数个适用于常见数据处理应用场景的扩展库。这些库通常嵌入在 API 中，且并不完全独立于其它 API。它们也因此可以受益于 API 的所有特性，并与其他库集成。
```
Comment (klion26) [inline flink-applications.zh.md]:
```suggestion
Flink 具有数个适用于常见数据处理应用场景的扩展库。这些库通常嵌入在 API 中，且并不完全独立于其它 API。它们也因此可以受益于 API 的所有特性，并与其他库集成。
```

File: flink-applications.zh.md [hunk 15]
```diff
@@ -133,59 +134,59 @@ public static class StartEndDuration
...
-  
-* **[Gelly](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/index.html)**: Gelly is a library for scalable graph processing and analysis. Gelly is implemented on top of and integrated with the DataSet API. Hence, it benefits from its scalable and robust operators. Gelly features [built-in algorithms](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/library_methods.html), such as label propagation, triangle enumeration, and page rank, but provides also a [Graph API](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/graph_api.html) that eases the implementation of custom graph algorithms.
+* **[DataSet API](https://ci.apache.org/projects/flink/flink-docs-stable/dev/batch/index.html)**：DataSet API 是 Flink 用于批处理应用程序的核心 API。DataSet API 的原语包括*map*、*reduce*、*(outer) join*、*co-group* 以及 *iterate*。所有算子都由算法和数据结构支持，这些算法和数据结构对内存中的序列化数据进行操作，如果数据大小超过预留内存则溢出到磁盘。Flink 的 DataSet API 的数据处理算法借鉴了传统数据库算法的实现，例如混合散列连接（hybrid hash-join）和外部归并排序（external merge-sort）。
```
Comment (klion26) [inline flink-applications.zh.md]: `所有算子都由算法和数据结构支持，这些算法和数据结构对内存中的序列化数据进行操作，如果数据大小超过预留内存则溢出到磁盘` may be improved.

File: flink-applications.zh.md [hunk 16]
```diff
@@ -133,59 +134,59 @@ public static class StartEndDuration
...
+
+* **[Gelly](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/index.html)**: Gelly 是一个可扩展的图形处理和分析库。Gelly 是在 DataSet API 之上实现的，并与 DataSet API 集成。因此，它能够受益于其可扩展且健壮的操作符。Gelly 提供了[内置算法](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/library_methods.html)，如label propagation、triangle enumeration 和 page rank 算法，也提供了一个简化自定义图算法实现的 [Graph API](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/graph_api.html)。
```
Comment (klion26) [inline flink-applications.zh.md]:
```suggestion
* **[Gelly](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/index.html)**: Gelly 是一个可扩展的图形处理和分析库。Gelly 是在 DataSet API 之上实现的，并与 DataSet API 集成。因此，它能够受益于其可扩展且健壮的操作符。Gelly 提供了[内置算法](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/library_methods.html)，如 label propagation、triangle enumeration 和 page rank 算法，也提供了一个简化自定义图算法实现的 [Graph API](https://ci.apache.org/projects/flink/flink-docs-stable/dev/libs/gelly/graph_api.html)。
```

File: flink-applications.zh.md [hunk 17]
```diff
@@ -16,70 +16,71 @@ title: "Apache Flink 是什么？"
...
+* **插件化的State Backend**：State Backend 负责管理应用程序状态，并在需要的时候进行 checkpoint。Flink 支持多种 state backend，可以将状态存在内存或者 [RocksDB](https://rocksdb.org/)。RocksDB 是一种高效的嵌入式、持久化键值存储引擎。Flink 也支持插件式的自定义 state backend 进行状态存储。
+* **精确一次语义**：Flink 的 checkpoint 和故障恢复算法保证了故障发生后应用状态的一致性。因此，Flink 能够在应用程序发生故障时，对应用程序透明，不造成正确性的影响。
+* **超大数据量状态**：Flink 能够利用其异步以及增量式的 checkpoint 算法，存储数 TB 级别的应用状态。
+* **可弹性伸缩的应用**：Flink 能够通过在更多或更少的工作节点上对状态进行重新分布，支持有状态应用的分布式的横向伸缩。
 
-### Time
+### 时间
 
-Time is another important ingredient of streaming applications. Most event streams have inherent time semantics because each event is produced at a specific point in time. Moreover, many common stream computations are based on time, such as windows aggregations, sessionization, pattern detection, and time-based joins. An important aspect of stream processing is how an application measures time, i.e., the difference of event-time and processing-time.
+时间是流处理应用另一个重要的组成部分。因为事件总是在特定时间点发生，所以大多数的事件流都拥有事件本身所固有的时间语义。进一步而言，许多常见的流计算都基于时间语义，例如窗口聚合、会话计算、模式检测和基于时间的 join。流处理的一个重要方面是应用程序如何处理时间，即事件时间（event-time）和处理时间（processing-time）的区别。
```
[the remark this answers is not in the document]
Comment (crazyzhou) [reply flink-applications.zh.md]: I asked whether we should translate `event-time` to `事件时间` in the google doc `Flink Translation Specifications` and @wuchong answered yes. I think it's ok with `处理时间`, maybe I can improve the `应用程序如何处理时间` to make it not as duplicated terms.

File: flink-applications.zh.md [hunk 18]
```diff
@@ -16,70 +16,71 @@ title: "Apache Flink 是什么？"
...
-* **Watermark Support**: Flink employs watermarks to reason about time in event-time applications. Watermarks are also a flexible mechanism to trade-off the latency and completeness of results.
-* **Late Data Handling**: When processing streams in event-time mode with watermarks, it can happen that a computation has been completed before all associated events have arrived. Such events are called late events. Flink features multiple options to handle late events, such as rerouting them via side outputs and updating previously completed results.
-* **Processing-time Mode**: In addition to its event-time mode, Flink also supports processing-time semantics which performs computations as triggered by the wall-clock time of the processing machine. The processing-time mode can be suitable for certain applications with strict low-latency requirements that can tolerate approximate results.
+* **事件时间模式**：使用事件时间语义的流处理应用根据事件本身自带的时间戳进行结果的计算。因此，无论处理的是历史记录的事件还是实时的事件，事件时间模式的处理总能保证结果的准确性和一致性。
+* **Watermark支持**：Flink 引入了 watermark 的概念，用以推断事件时间应用的时间。Watermark 也是一种平衡处理延时和完整性的灵活机制。
```
[the remark this answers is not in the document]
Comment (crazyzhou) [reply flink-applications.zh.md]: `reason about time` is hard to translate directly to keep its meaning. Can I change it into `衡量事件时间进展` as my understanding of watermark?

File: flink-applications.zh.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (wuchong) [reply flink-applications.zh.md]: What about change `应用程序如何处理时间` to `应用程序如何衡量时间` ?

File: flink-applications.zh.md (hunk 7 above)
Comment (crazyzhou) [reply flink-applications.zh.md]: I think `timer` should be translated as it's lowercased as a general meaning and not that anbiguous. Agree that `定时器` is better.

File: flink-applications.zh.md (hunk 15 above)
Comment (crazyzhou) [reply flink-applications.zh.md]: Any suggestions?

**Comment by wuchong** [comment]
Hi @crazyzhou  @klion26 , what's the status of the pull request now ?

**Comment by crazyzhou** [comment]
I think I've finished all the fixes, just need a final check from @wuchong  or @klion26 .

**Comment by klion26** [comment]
LGTM +1

**Comment by wuchong** [comment]
Thanks @klion26 ,  I will merge it.

**Pull request closed without merging**